### ISSUE TITLE
Compilation failure: zlib/zlib.h: No such file or directory

### ISSUE THREAD

**Issue by Szunti** [freeorion/freeorion#373]
Tried to compile master (57819202):

```
[  0%] Building CXX object CMakeFiles/freeorioncommon.dir/util/DependencyVersions.cpp.o
repos/freeorion/util/DependencyVersions.cpp:6:23: fatal error: zlib/zlib.h: No such file or directory
compilation terminated.
CMakeFiles/free[  0%] Building CXX object CMakeFiles/freeorioncommon.dir/util/DependencyVersions.cpp.o
/home/zoli/repos/freeorion/util/DependencyVersions.cpp:6:23: fatal error: zlib/zlib.h: No such file or directory
compilation terminated.
CMakeFiles/freeorioncommon.dir/build.make:878: recipe for target 'CMakeFiles/freeorioncommon.dir/util/DependencyVersions.cpp.o' failed
make[2]: *** [CMakeFiles/freeorioncommon.dir/util/DependencyVersions.cpp.o] Error 1
CMakeFiles/Makefile2:67: recipe for target 'CMakeFiles/freeorioncommon.dir/all' failed
make[1]: *** [CMakeFiles/freeorioncommon.dir/all] Error 2
Makefile:149: recipe for target 'all' failed
make: *** [all] Error 2
orioncommon.dir/build.make:878: recipe for target 'CMakeFiles/freeorioncommon.dir/util/DependencyVersions.cpp.o' failed
make[2]: *** [CMakeFiles/freeorioncommon.dir/util/DependencyVersions.cpp.o] Error 1
CMakeFiles/Makefile2:67: recipe for target 'CMakeFiles/freeorioncommon.dir/all' failed
make[1]: *** [CMakeFiles/freeorioncommon.dir/all] Error 2
Makefile:149: recipe for target 'all' failed
make: *** [all] Error 2
```

I use Archlinux, and have zlib installed. Tried make clean, rerun ccmake and compile, but still got this.

**Comment by geoffthemedio** [comment]
Try removing zlib/ before zlib.h from the #include in DependencyVersions.cpp

Probably similar adjustments would be needed for other includes in that file... possibly most of them need commenting out until someone who knows how to fix up the CMake settings to find all those dependency include directories does so.

**Comment by Szunti** [comment]
I could advance with removing zlib/ and png/, now I need to do something with ft2build.h, I don't see this header in freetype2(2.6.1) files.

Sorry, I was wrong, I have that file.

**Comment by geoffthemedio** [comment]
I'd hope so, as GG's Font.cpp has the same style of include for FreeType stuff...

`#include <ft2build.h>`
`#include FT_FREETYPE_H`

(The Windows SDK has ft2build.h in the main freetype include directory.)

**Comment by geoffthemedio** [comment]
Does this set of #includes work?

```
#include "Version.h"

#include "Logger.h"

#include <zlib.h>
#include <SDL_version.h>
#include <png.h>
#include <patchlevel.h>

#include <ft2build.h>
#include FT_FREETYPE_H

#include <boost/version.hpp>
#include <sstream>
```

**Comment by Szunti** [comment]
I need the SDL2/ before SDL_version and ft2build.h is not found

**Comment by Vezzra** [comment]
I've got similar issues on OSX with png.h. Removing the png/ before png.h seems to solve that.

However, I need to update the OSX SDK now, as it doesn't include the zlib headers (which hasn't been an issue until now, as the zlib library has only been required by other, precompiled libraries). I'll go ahead assuming that `#include <zlib.h>` is going to be the form used (as that seems to be what works on Linux too).

**Comment by Szunti** [comment]
Somehow the -I switches (that cmake should add) are missing from the command run by make.

**Comment by Vezzra** [comment]
> I need the SDL2/ before SDL_version

Yep, same on OSX so far, but...

>  and ft2build.h is not found

...this doesn't make sense. As Geoff already pointed out, GG's Font.h has that same include in exact that style since forever (AFAIK). If that didn't give you any problems till now, why should exactly the same include suddenly create issues when used in DependencyVersions.cpp?

I've got no problems with that on OSX.

**Comment by Szunti** [comment]
Because GG has an additional CMakeLists.txt that adds include for freetype, png and others

**Comment by Vezzra** [comment]
Ah, I see. Well, that means we need to update the respective CMakeLists.txt file(s) accordingly I guess...

**Comment by geoffthemedio** [comment]
If the OSX SDK doesn't include the zlib headers, then how has Message.cpp been including zlib.h for years?

https://github.com/freeorion/freeorion/blame/master/network/Message.cpp#L29

**Comment by Szunti** [comment]
The server doesn't need SDL, png, freetype. It doesn't make sense to print their version in 57819202.
I think, you shouldn't add additional dependencies to common just to print their version.

**Comment by geoffthemedio** [comment]
It doesn't add the dependency in the sense of requiring the shared library, it just needs to extract the version info from the header. I specifically excluded some dependencies to avoid actually adding them in that sense.

**Comment by Szunti** [comment]
So you need the header, which would almost certainly mean you need to have the libraries too. You can't really install them seperately. 

Maybe you should put these functions into a header, where you can check for USE_ -defines made by CMAKE to check only for libraries the code really use.

**Comment by Vezzra** [comment]
> If the OSX SDK doesn't include the zlib headers, then how has Message.cpp been including zlib.h for years?

Well, zlib actually is provided by OSX, and its header files by Apples system SDKs, so I never noticed this glitch. The compiler just falls back on the zlib.h of the system SDK. Apparently, this never caused any issues, although the zlib shipping with FOs OSX SDK is version 1.2.7, while the version provided by OSX is 1.2.5.

This is not exactly ideal, so I've to update the SDK now anyway. I'll wait however for this issue here to be ironed out first, so it'll be clear what the final style/form of the zlib/png/whatever includes are going to be and I only have to do this update once.

**Comment by Vezzra** [comment]
@Szunti, I've [pushed a branch](https://github.com/Vezzra/freeorion/commits/cmake-update) with [updated CMakeLists.txt](https://github.com/Vezzra/freeorion/commit/ff849e56e5005efaa5a3c5887c1e6c09b18a28df) to [my fork of the freeorion repo](https://github.com/Vezzra/freeorion.git). Does this setup work for you?

**Comment by Szunti** [comment]
I don't think it's good to add freetype as dependency, when it's easy to solve it cleanly with moving those functions to a header. But if you choose that way, you should also require sdl2 and maybe python?

@Vezzra DependencyVersions.cpp compiles with your branch, the full compilation will take some time.

**Comment by geoffthemedio** [comment]
You already need all these libraries to build all the parts of FreeOrion... so what's the problem?

You'd have a point if someone just wanted to compile the server and not the AI or human clients, but how often is that the case...?

**Comment by Vezzra** [comment]
> I don't think it's good to add freetype as dependency, when it's easy to solve it cleanly with moving those functions to a header.

Well, lets see what @geoffthemedio comes up with. I assume he's going to work or probably already is working on this.

> But if you choose that way, you should also require sdl2 and maybe python?

You mean, add both requirements to both the server and the human client? Because SDL2 of course is already a requirement of the client (well, not directly, but via GG AFAIK), and Python already a requirement of the server (server-side universe generation is done by Python scripts).

> DependencyVersions.cpp compiles with your branch, the full compilation will take some time.

Ok, thanks. So at least now we know how to fix CMakeLists.txt (just in case we decide to add the extra dependencies).

**Comment by Vezzra** [comment]
> You already need all these libraries to build all the parts of FreeOrion... so what's the problem?

Hm, I guess @Szunti is talking about adding dependencies to executables that don't really require it. The server doesn't need SDL, freetype and the PNG library, the human client doesn't need Python (so far).

However, as far as I can see, just including the header files does not actually add the requirement to link against the respective libraries. So the question is what the problem is supposed to be here - @Szunti?

**Comment by Vezzra** [comment]
Anyway, FYI, this is the set of includes that works on OSX, when I just add the zlib header files to the SDK without doing any further adjustments:

```
#include "Version.h"

#include "Logger.h"

#include <SDL2/SDL_version.h>
#include <zlib.h>
#include <png.h>
#include <python2.7/patchlevel.h>

#include <ft2build.h>
#include FT_FREETYPE_H

#include <boost/version.hpp>
#include <sstream>
```

**Comment by Szunti** [comment]
But if you don't link to a lib, why would you need the headers? The version of libraries you are not using are irrelevant. 

There is a layer already that manages this problem. You can use cmake to configure which libraries to use, and can use that information in source files. Now you are removing this, and also some boundaries between server and client. Also the possibility for optional libraries. (or they just won't be reported in logs).

And I don't see the reason. You can easily add to LogDependencyVersions() ifdef-s to print only the used libraries, and that would solve all the problems.

Building only the server is a sane thing for an online multiplayer game. There are dedicated servers for quake, openra, spring, freeciv for example.

**Comment by geoffthemedio** [comment]
The headers are used to get the versions of the various dependencies that are used by some parts of FreeOrion.

Not all builds use CMake, so CMake-specific defines can't be checked for in all cases.

**Comment by geoffthemedio** [comment]
If you can provide a version of DependencyVersions.cpp that uses #ifdef or similar checks CMake-related defines in order to stop building of some functions (and calling them later in the file) when various headers aren't available, or perhaps if certain projects that use those dependencies aren't being built, then please do so... I can't write anything of that sort myself as I don't build using CMake.

**Comment by Vezzra** [comment]
To expand a bit on @geoffthemedio's replies:

> But if you don't link to a lib, why would you need the headers? The version of libraries you are not using are irrelevant.

Correct, but logging only the versions of libraries a particular binary uses would require to either maintain separate DependencyVersion.cpp source files for each binary in question, or add respective `#ifdef` statements to DependencyVersion.cpp. @geoffthemedio's approach is just a bit easier.

> And I don't see the reason. You can easily add to LogDependencyVersions() ifdef-s to print only the used libraries, and that would solve all the problems.

That should work well enough though, without having to touch the CMake files at all if I understand correctly (the #idefs would just have to check which binary is being build, right?). Any reason why we shouldn't do it this way? @geoffthemedio?

> Building only the server is a sane thing for an online multiplayer game. There are dedicated servers for quake, openra, spring, freeciv for example.

FO is still far from providing anything like that. So that's not really relevant in this case (at least for now).

**Comment by geoffthemedio** [comment]
You need to #include headers before you can check whether things are defined by them. If those headers don't exist, that step will fail. So you can't just check if PY_VERSION is defined before defining/declaring/using functions that depend on it.

As far as I know, you can't check which binary is being built, because Common is not a binary, or is a shared library that is used by several different binaries, so none of the executable/binary-specific defines (eg. FREEORION_BUILD_HUMAN) will be defined in that code. You'd be better off having a separate version of DependencyVersions.cpp for each binary for that method, which I wanted to avoid. Since #including the headers doesn't in of itself add a shared-object/static-library/dll build-time or run-time dependency, this seemed fine...

**Comment by Szunti** [comment]
I just read cmake docs to know how it works, and only GG knows about optional dependencies.

But do these version strings useful anyway? Most people will use prebuilt packages, and the printed versions are the ones on the machine which built the package, not the local ones.

**Comment by Vezzra** [comment]
> As far as I know, you can't check which binary is being built, because Common is not a binary,

Argh, right of course, DependencyVersions.cpp is compiled into Common. That rules out any solution with #ifdefs AFAIK. Leaving only separate versions of DependencyVersions.cpp for each binary, which would be a bit of overkill I guess...

**Comment by Szunti** [comment]
You can put the code into a header to solve that.

**Comment by Vezzra** [comment]
> But does these version strings useful anyway? Most people will use prebuilt packages, and the printed versions are the ones on the machine which built the package, not the local ones.

On Windows and OSX it does make plenty of sense, as these are the versions of dependencies that are shipped with the FO package on these platforms, meaning, the versions of the libraries actually used by the binaries.

However, I guess you might have a bit of a point regarding Linux, at least on distros that provide precompiled packages in their repos (instead of source packages that get locally built). There it probably isn't quite as useful, as it will only tell which version of the libraries have been used on the system on which the precomiled packages had been built.

But that might still be useful information even in this case, so yes, I think these version strings are useful.

**Comment by Vezzra** [comment]
> You can put the code into a header to solve that.

You mean, turn DependencyVersions.cpp into DependencyVersions.h (more or less)?

**Comment by Vezzra** [comment]
@geoffthemedio: The above discussion aside, can we at least already agree on the style/form of the #include statements in DependencyVersions.cpp? Just so I can update the OSX SDK correctly...

**Comment by Szunti** [comment]
They should have the form that works with CMake (or what you use to get the include flags for compiler). You should probably use what is used in GG I think.

**Comment by Vezzra** [comment]
As far as I can tell, that should be the set of includes I posted above which works on OSX. @geoffthemedio, can we agree on this set? If yes, I'd go ahead and push it to master.

**Comment by geoffthemedio** [comment]
If a certain set of includes works for you, it's probably fine. I should be able to adjust the MSVC include paths to compensate. But not right now... in a hurry...

**Comment by Szunti** [comment]
I have to admit I couldn't find a nice solution. It either makes it harder to add new libraries (because not just one line needed in CMakeLists.txt and it also requires a line in a template config header file that needs to be included before these dependency logging functions), or need to generate source code.

**Comment by Mitten-O** [comment]
I put up what fixed this for me (linux) as #374 .

**Comment by Vezzra** [comment]
I've updated the MSVC 2013 project file (commit dad9de2d02129e829ec1e7db4b5cc3187551b6f8), building on Windows now works with the changes Mitten.O put in to fix building on Linux.

As building on OSX works too, I consider this issue as solved for now.
